File: frontend/src/services/UserService.ts
import axios from "axios";
import { Address } from "../types/Address";
import { Payment } from "../types/Payment";
import { User } from "../types/User";
import { API_URL } from "../constants";
import { ValidateResponse } from "../types/ValidateResponse";

export class UserService {
	// Get either a user or a registered user by ID
	async getUserOrRegisteredUserById(userId: number): Promise<User> {
		try {
			const response = await axios.get(`${API_URL}/users/${userId}`);
			console.log("User response:", response.data);
			const payment = await this.getPaymentById(response.data.paymentId);
			const address = await this.getAddressById(response.data.addressId);
			const email = await this.getEmailByUserId(response.data.id);
			const user: User = {
				id: response.data.id,
				email: email,
				payment: payment,
				address: address,
			};
			return user;
		} catch (error) {
			throw error;
		}
	}

	// Create a user with just an email
	async createUser(email: string): Promise<User> {
		try {
			const response = await axios.post(`${API_URL}/users/create`, null, {
				params: { email },
			});
			return response.data;
		} catch (error) {
			throw error;
		}
	}

	// Register a user with address, payment, and password details
	async registerUser(
		email: string,
		street: string,
		city: string,
		state: string,
		zipCode: string,
		cardNumber: string,
		expiry: string,
		cvv: string,
		password: string
	): Promise<User> {
		try {
			const response = await axios.post(
				`${API_URL}/users/register`,
				{
					email,
					street,
					city,
					state,
					zipCode,
					cardNumber,
					expiry,
					cvv,
					password,
				},
				{
					headers: {
						"Content-Type": "application/json",
					},
				}
			);
			return response.data;
		} catch (error) {
			throw error;
		}
	}

	// Retrieve user ID by email
	static async getUserIdByEmail(email: string): Promise<number> {
		try {
			const response = await axios.get(`${API_URL}/users/id`, {
				params: { email },
			});
			return response.data;
		} catch (error) {
			throw error;
		}
	}
	// Retrieve email by user ID
	async getEmailByUserId(userId: number): Promise<string> {
		try {
			const response = await axios.get(`${API_URL}/users/${userId}/email`);
			return response.data;
		} catch (error) {
			throw error;
		}
	}

	// Validate password
	async validatePassword(
		email: string,
		password: string
	): Promise<User | null> {
		try {
			const response = await axios.post(`${API_URL}/users/validate`, null, {
				params: { email, password },
			});

			const validateResponse: ValidateResponse = response.data;

			if (validateResponse.valid && validateResponse.userId) {
				return this.getUserOrRegisteredUserById(validateResponse.userId);
			}
			return null;
		} catch (error) {
			throw error;
		}
	}

	// Get address by ID
	async getAddressById(addressId: number): Promise<Address> {
		try {
			const response = await axios.get(`${API_URL}/users/address/${addressId}`);
			return response.data;
		} catch (error) {
			throw error;
		}
	}

	// Get payment by ID
	async getPaymentById(paymentId: number): Promise<Payment> {
		try {
			const response = await axios.get(`${API_URL}/users/payment/${paymentId}`);
			return response.data;
		} catch (error) {
			throw error;
		}
	}
}

File: backend/src/main/java/com/AcmeBuddy/backend/services/UserService.java
package com.AcmeBuddy.backend.services;

import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import com.AcmeBuddy.backend.entities.Address;
import com.AcmeBuddy.backend.entities.Payment;
import com.AcmeBuddy.backend.entities.RegisteredUser;
import com.AcmeBuddy.backend.entities.User;
import com.AcmeBuddy.backend.repositories.IAddressRepository;
import com.AcmeBuddy.backend.repositories.IPaymentRepository;
import com.AcmeBuddy.backend.repositories.IRegisteredUserRepository;
import com.AcmeBuddy.backend.repositories.IUserRepository;

import jakarta.transaction.Transactional;

@Service
public class UserService {

    private final IUserRepository userRepository;
    private final IRegisteredUserRepository registeredUserRepository;
    private final IPaymentRepository paymentRepository;
    private final IAddressRepository addressRepository;
    private final PasswordEncoder passwordEncoder;

    @Autowired
    public UserService(IUserRepository userRepository, IRegisteredUserRepository registeredUserRepository,
            IPaymentRepository paymentRepository, IAddressRepository addressRepository) {
        this.userRepository = userRepository;
        this.registeredUserRepository = registeredUserRepository;
        this.paymentRepository = paymentRepository;
        this.addressRepository = addressRepository;
        this.passwordEncoder = new BCryptPasswordEncoder();
    }

    // Create a user with just an email
    public User createUser(String email) {
        if (userRepository.findByEmail(email).isPresent()) {
            throw new IllegalArgumentException("User with this email already exists.");
        }
        User user = new User(email);
        userRepository.save(user);
        return user;
    }

    @Transactional
    public RegisteredUser registerUser(String email, String street, String city, String state, String zipCode,
            String cardNumber, String expiry, String cvv, String password) {
        User user = userRepository.findByEmail(email).orElse(null);

        if (user == null) {
            user = new User(email);
            userRepository.save(user);
        }

        if (registeredUserRepository.findById((long) user.getId()).isPresent()) {
            throw new IllegalArgumentException("This user is already registered.");
        }

        Address address = new Address(street, city, state, zipCode);
        addressRepository.save(address);

        Payment payment = new Payment(cardNumber, expiry, cvv);
        paymentRepository.save(payment);

        String encodedPassword = passwordEncoder.encode(password);
        RegisteredUser registeredUser = new RegisteredUser(user, payment.getId(), address.getId(), encodedPassword);

        return registeredUserRepository.save(registeredUser);
    }

    // Retrieve user ID by email
    public Optional<Integer> getUserIdByEmail(String email) {
        return userRepository.findByEmail(email).map(User::getId);
    }

    public Optional<RegisteredUser> validatePassword(String email, String password) {
        System.out.println("Fetching user with email: " + email);
        Optional<RegisteredUser> registeredUserOpt = registeredUserRepository.findByUserEmail(email);
        if (registeredUserOpt.isPresent()) {
            RegisteredUser registeredUser = registeredUserOpt.get();
            System.out.println("Password comparison result: " + passwordEncoder.matches(password, registeredUser.getPassword()));

            if (passwordEncoder.matches(password, registeredUser.getPassword())) {
                return Optional.of(registeredUser);
            }
        }

        return Optional.empty();
    }

    // Get address by ID
    public Optional<Address> getAddressById(Long addressId) {
        return addressRepository.findById(addressId);
    }

    // Get payment by ID
    public Optional<Payment> getPaymentById(Long paymentId) {
        return paymentRepository.findById(paymentId);
    }

    public Optional<Object> getUserOrRegisteredUserById(Long userId) {
        Optional<RegisteredUser> registeredUser = registeredUserRepository.findById(userId);
        if (registeredUser.isPresent()) {
            return Optional.of(registeredUser.get());
        }
        return userRepository.findById(userId).map(user -> (Object) user);
    }

    public Optional<String> getEmailByUserId(Long userId) {
        Optional<User> user = registeredUserRepository.findUserWithRegisteredUserById(userId);
        if (user.isPresent()) {
            return Optional.of(user.get().getEmail());
        }
        return Optional.empty();
    }
}
